### PR TITLE
Story(CCLS-2168) Pre-populate means and merits assessments

### PR DESCRIPTION
**Pull request by PhilDigitalJustice**
includes:
- initial implementation of the assessment controller
- Removal of redundant assessment data
- pre pop of assessment data
- mapping of an application and client data into an assessment.
- new utility classes to supports assessment, opponent, proceeding and application classes.
- new css and js files for the assessment screen (not yet fully implemented)

### CODE REVIEW

File: src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java:1]: Please can these be compared to the files:

com.legalservices.laa.ccms.config.MappingConfig In PUI

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+
+
+  /**
+   * Converts proceeding data to a list of {@link AssessmentAttributeDetail} objects.
+   *
+   * @param proceeding the proceeding detail containing various attributes
+   * @return a list of assessment attribute details for the proceeding
+   */
+  @Named("proceedingToAttributeList")
+  default List<AssessmentAttributeDetail> proceedingToAttributeList(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:276]: Please can these be compared to the files:

com.legalservices.laa.ccms.config.MappingConfig In PUI

see - public CcmsInstancePrepopulator foreachProceeding() {

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+  }
+
+  /**
+   * Converts opponent data to a list of {@link AssessmentAttributeDetail} objects.
+   *
+   * @param opponentContext the opponent mapping context containing opponent details
+   * @return a list of assessment attribute details for the opponent
+   */
+  @Named("opponentToAttributeList")
+  default List<AssessmentAttributeDetail> opponentToAttributeList(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:351]: Please can these be compared to the files:

com.legalservices.laa.ccms.config.MappingConfig In PUI

see - public CcmsInstancePrepopulator foreachOpponent() {

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+  }
+
+  /**
+   * Converts global assessment data to a list of {@link AssessmentAttributeDetail} objects.
+   *
+   * @param context the assessment mapping context containing application, client, and user details
+   * @return a list of assessment attribute details
+   */
+  @Named("globalToAttributeList")
+  default List<AssessmentAttributeDetail> globalToAttributeList(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:379]: Please can these be compared to the files:

com.legalservices.laa.ccms.config.MappingConfig In PUI

see - public List<Prepopulator> sharedPrepopulationMappingList() {

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 4]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
 
   /**
-   * Retrieves the most recent date on which any key data was modified within the given
-   * application. This includes checks across proceedings and opponents involved in the
+   * Cleans up data by deleting redundant opponents and proceedings from the assessment.
+   *
+   * @param assessment the assessment detail to clean up
+   * @param application the application detail for reference
+   */
+  public void cleanupData(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:493]: Please can these be compared to the files:

com.ezgov.legalservices.ccms.utils.IframeHelper In PUI

The logic should be comparable to:
public boolean cleanupData(OpaSession opaSession, Case app) {

It has been cleaned up and duplication removed where possible

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java [hunk 5]
```diff
@@ -76,4 +94,113 @@ public String assessmentRemovePost(
...
+   *
+   * @param assessment the assessment to remove.
+   * @param invokedFrom the page from which the assessment is being invoked.
+   * @param user the user making the request.
+   * @param session the session in which the request is being made.
+   * @param request the http servlet request being made.
+   * @return the view that displays the OPA assessment.
+   */
+  @GetMapping("/assessments")
+  public String assessmentGet(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:108]: Please can these be compared to the files:

com.legalservices.laa.ccms.controller.iframeHoldingController In PUI

@RequestMapping(value = {"/renderOWD","/renderOWD.do"}, method = RequestMethod.GET)
  @Transactional
  public String editIFrameHolding(Map<String, Object> model,
      HttpServletRequest request,
      @RequestParam("invokedFrom") String invokedFrom,
      @RequestParam("assessmentType") String assessmentType,
      @RequestParam("ezgov_private_hiddenData_uniqueId") String uniqueID) {
      
Logic has been replicated up to the point of creating the means and merits data in the opa session tables. 
Anything todo with amendments or case data has been left as a todo as this implementation is only for initial applications.

The controller will redisplay the summary screen, still to implement and integrate fully with OPA to display the assessments.

File: src/main/java/uk/gov/laa/ccms/caab/client/AssessmentApiClient.java [hunk 6]
```diff
@@ -85,6 +88,81 @@ public Mono<Void> deleteAssessments(
...
+  }
+
+  /**
+   * Deletes a checkpoint for a given assessment.
+   *
+   * @param assessmentId the ID of the assessment
+   * @param userLoginId the login ID of the user
+   * @return a Mono that completes when the deletion is finished
+   */
+  public Mono<Void> deleteAssessmentCheckpoint(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/client/AssessmentApiClient.java:149]: Remove this method

File: src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java [hunk 7]
```diff
@@ -7,21 +7,81 @@
...
    *
-   * @param attribute the string representation of the assessment attribute
+   * @param name the name of the assessment attribute
+   * @param type the type of the assessment attribute
+   * @param prepopulated whether the attribute is prepopulated
+   * @param asked whether the attribute is asked
    */
-  AssessmentAttribute(final String attribute) {
-    this.attribute = attribute;
+  AssessmentAttribute(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java:77]: Remove name, and use the enum entry value

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java [hunk 8]
```diff
@@ -76,4 +94,113 @@ public String assessmentRemovePost(
...
+    if (!assessmentRulebase.isFinancialAssessment()) {
+      final AssessmentDetails assessmentDetails = assessmentService.getAssessments(
+          List.of(assessmentRulebase.getPrePopAssessmentName()),
+          String.valueOf(user.getProvider().getId()),
+          application.getCaseReferenceNumber(),
+          null).block();
+
+      for (final AssessmentDetail prepopAssessment : assessmentDetails.getContent()) {
+
+        if (prepopAssessment.getName()
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:147]: Remove this check as the assessment service/ api filter will do this

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java [hunk 9]
```diff
@@ -76,4 +94,113 @@ public String assessmentRemovePost(
...
+    //delete opponents and proceedings that have been removed from the application
+    // if not a financial assessment type (aka means or merits)
+    if (!assessmentRulebase.isFinancialAssessment()) {
+      final AssessmentDetails assessmentDetails = assessmentService.getAssessments(
+          List.of(assessmentRulebase.getPrePopAssessmentName()),
+          String.valueOf(user.getProvider().getId()),
+          application.getCaseReferenceNumber(),
+          null).block();
+
+      for (final AssessmentDetail prepopAssessment : assessmentDetails.getContent()) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:145]: This should be find first, remove the break then

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java [hunk 10]
```diff
@@ -76,4 +94,113 @@ public String assessmentRemovePost(
...
+    //get full client details from soa
+    // If user has not modified client details,
+    // it is necessary to do an EBS call to get a fully populated client
+    //lets leave this for now.
+    final ClientDetail client = clientService.getClient(
+        application.getClient().getReference(),
+        user.getLoginId(),
+        user.getUserType()).block();
+
+    if (assessment.equalsIgnoreCase("means") || assessment.equalsIgnoreCase("merits")) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:174]: change to is non financial assessment

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java [hunk 11]
```diff
@@ -76,4 +94,113 @@ public String assessmentRemovePost(
...
+      final String contextToken = contextSecurityUtil.createHubContext(
+          application.getCaseReferenceNumber(), assessmentRulebase.getId(), user.getUsername(),
+          user.getProvider().getId().longValue(), request.getSession().getId(), invokedFrom,
+          "1234567890");
+
+      //start opa assessment
+      assessmentService.startAssessment(
+          application,
+          assessmentRulebase,
+          application.getCaseReferenceNumber(),
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:187]: remove case reference, we already have it in the application

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 12]
```diff
@@ -0,0 +1,497 @@
...
+        getAssessmentEntityType(context.getAssessment(), GLOBAL);
+
+    if (existingEntityType != null) {
+      return existingEntityType;
+    } else {
+
+      //opponent relationships
+      final List<AssessmentRelationshipTargetDetail> opponentRelationshipTargets =
+          new ArrayList<>();
+      for (final OpponentDetail opponent : context.getApplication().getOpponents()) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:153]: Convert to lambda stream

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:1]: Think about adding null checks to all the attributes

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 13]
```diff
@@ -0,0 +1,497 @@
...
+   * Converts proceeding data to a list of {@link AssessmentAttributeDetail} objects.
+   *
+   * @param proceeding the proceeding detail containing various attributes
+   * @return a list of assessment attribute details for the proceeding
+   */
+  @Named("proceedingToAttributeList")
+  default List<AssessmentAttributeDetail> proceedingToAttributeList(
+      final ProceedingDetail proceeding) {
+    return List.of(
+        toAttribute(CLIENT_INVOLVEMENT_TYPE,
```
Comment (porritta) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:279]: toClientInvolvementTypeAttribute(proceeding);

@Mapping(target="", source="clientInvolvement.id")
AssessmentAttribute toClientInvolvementTypeAttribute(ProceedingDetail proceeding);

File: src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentMappingContext.java [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Context class for mapping assessment details.
+ */
+@Builder
+@Getter
+@Setter
+public class AssessmentMappingContext {
+
+  private AssessmentDetail assessment;
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentMappingContext.java:21]: Add java doc to the variables

File: src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentOpponentMappingContext.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+
+
+/**
+ * Context class for mapping an assessment opponent details.
+ */
+@Builder
+@Getter
+public class AssessmentOpponentMappingContext {
+
+  private final OpponentDetail opponent;
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentOpponentMappingContext.java:16]: java doc the variables

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 16]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
-      }
+
+      final CommonLookupValueDetail titleCommonLookup = lookupService.getCommonValue(
+              COMMON_VALUE_CONTACT_TITLE, opponent.getTitle())
+          .map(commonLookupValueDetail -> commonLookupValueDetail
+              .orElse(new CommonLookupValueDetail()
+                  .code(opponent.getTitle())
+                  .description(opponent.getTitle())))
+          .blockOptional()
+          .orElseThrow();
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:853]: Throw caab api exception

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 17]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
+
+    updateCostLimitIfMeritsAssessment(assessmentRulebase, application);
+
+    if (prepopulateFromEbs) {
+      //todo - later implementation in future story
+    }
+
+    //if means and merits:
+    if (!assessmentRulebase.isFinancialAssessment()) {
+      if (isOpaSessionClean(prepopAssessment)
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:817]: Amend to isAseesssmentClean

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 18]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
-   * @return the matching assessment entity type detail, or null if not found
+   * @param application the application details containing proceedings
+   * @param assessment the assessment details to check against
+   * @return true if a proceeding in the application does not exist in the assessment or if the
+   *         scope of any proceeding has changed, otherwise false
    */
-  private AssessmentEntityTypeDetail getAssessmentEntityType(
-      final AssessmentDetail assessment,
-      final String entityTypeName) {
+  protected boolean isAppProceedingsExistInOpa(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:983]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 19]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
-   * @return the matching assessment entity detail, or null if not found
+   * @param application the application details containing opponents
+   * @param assessment the assessment details containing opponent entities
+   * @return true if the opponent counts do not match or if an opponent entity in the assessment
+   *         does not exist in the application, otherwise false
    */
-  private AssessmentEntityDetail getAssessmentEntity(
-      final AssessmentEntityTypeDetail assessmentEntityType,
-      final String entityName) {
+  protected boolean isOpponentCountMatchOpa(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:1028]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 20]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
-   * @return the matching assessment attribute detail, or null if not found
+   * @param application the application details containing opponents
+   * @param assessment the assessment details containing opponent entities
+   * @return true if an opponent entity in the assessment does not exist in the application,
+   *         otherwise false
    */
-  private AssessmentAttributeDetail getAssessmentAttribute(
-      final AssessmentEntityDetail assessmentEntity,
-      final String attributeName) {
+  protected boolean isOpaOpponentsMatchApplication(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:1047]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 21]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
 
+  /**
+   * Checks if any opponent in the application does not exist in the assessment's OPA session.
+   *
+   * @param application the application details containing opponents
+   * @param assessment the assessment details to check against
+   * @return true if an opponent in the application does not exist in the assessment, otherwise
+   *         false
+   */
+  protected boolean isApplicationMatchOpaOpponents(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:1083]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java [hunk 22]
```diff
@@ -447,137 +485,640 @@ private boolean differenceGreaterThanTenSecs(final Date keyChangeDate, final Dat
...
+  /**
+   * Checks if any proceeding entity in the assessment does not exist in the application's
+   * proceedings.
+   *
+   * @param application the application details containing proceedings
+   * @param assessment the assessment details containing proceeding entities
+   * @return true if a proceeding entity in the assessment does not exist in the application, or
+   *         if there is a mismatch with application proceedings, otherwise false
+   */
+  protected boolean isOpaProceedingsMatchApplication(
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:946]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+    }
+  }
+
+  /**
+   * Checks if the provider has a contract under devolved powers.
+   *
+   * @param application the application details
+   * @return true if the devolved powers contract flag starts with "yes", otherwise false
+   */
+  public static boolean getProviderHasContractOpaInput(final ApplicationDetail application) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java:40]: Rename method replace opa with Assessment, rename method to make sense

File: src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+    return false;
+  }
+
+  /**
+   * Checks if the application type is "EXCEPTIONAL_CASE_FUNDING".
+   *
+   * @param application the application details
+   * @return true if the application type is "EXCEPTIONAL_CASE_FUNDING", otherwise false
+   */
+  public static boolean getEcfFlagOpaInput(final ApplicationDetail application) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java:56]: Rename method replace opa with Assessment, rename method to make sense

File: src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java [hunk 25]
```diff
@@ -0,0 +1,127 @@
...
+        .equalsIgnoreCase(APP_TYPE_EXCEPTIONAL_CASE_FUNDING);
+  }
+
+  /**
+   * Checks if the LAR scope flag is set to true.
+   *
+   * @param application the application details
+   * @return true if the LAR scope flag is not null and true, otherwise false
+   */
+  public static boolean getLarScopeFlagOpaInput(final ApplicationDetail application) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java:69]: Rename method replace opa with Assessment, rename method to make sense

File: src/main/java/uk/gov/laa/ccms/caab/util/AssessmentUtil.java [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+        .orElse(null);
+  }
+
+  /**
+   * Checks if the OPA session for the given assessment detail is clean.
+   *
+   * @param assessment the assessment detail to check
+   * @return {@code true} if the session is clean, {@code false} otherwise
+   */
+  public static boolean isOpaSessionClean(final AssessmentDetail assessment) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/AssessmentUtil.java:135]: Rename method replace opa with Assessment

File: src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java [hunk 27]
```diff
@@ -0,0 +1,153 @@
...
+  public static final String ORGANISATION_OPPONENT = "Organisation";
+
+
+  /**
+   * Retrieves the OPA instance mapping ID for the given opponent detail.
+   *
+   * @param opponent the opponent detail
+   * @return the OPA instance mapping ID, prefixed if EBS ID is not available
+   */
+  public static String getOpaInstanceMappingId(final OpponentDetail opponent) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java:26]: Rename method replace opa with Assessment, rename method to make sense

File: src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java [hunk 28]
```diff
@@ -0,0 +1,153 @@
...
+    return opponent.getEbsId();
+  }
+
+  /**
+   * Determines the OPA opponent type based on the given type string.
+   *
+   * @param type the type string to check
+   * @return the opponent type, either as an organisation or a person
+   */
+  public static String getOpaOpponentType(final String type) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java:39]: Rename method replace opa with Assessment, rename method to make sense

File: src/main/java/uk/gov/laa/ccms/caab/util/ProceedingUtil.java [hunk 29]
```diff
@@ -0,0 +1,120 @@
...
+
+
+
+  /**
+   * Returns the OPA instance mapping ID for the given proceeding.
+   *
+   * @param proceeding the proceeding to get the OPA instance mapping ID for
+   * @return the OPA instance mapping ID
+   */
+  public static String getOpaInstanceMappingId(final ProceedingDetail proceeding) {
```
Comment (PhilDigitalJustice) [inline src/main/java/uk/gov/laa/ccms/caab/util/ProceedingUtil.java:26]: rename OPA to assessment

File: src/main/java/uk/gov/laa/ccms/caab/client/AssessmentApiClient.java (hunk 6 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/client/AssessmentApiClient.java:149]: Removed

File: src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java (hunk 7 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java:77]: Amended

Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/constants/assessment/AssessmentAttribute.java:77]: amended

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java (hunk 8 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:147]: Removed

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java (hunk 9 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:145]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java (hunk 10 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:174]: amended

File: src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java (hunk 11 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/controller/AssessmentController.java:187]: Removed

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java (hunk 12 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:153]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java [hunk 30]
```diff
@@ -0,0 +1,497 @@
...
+        opponentRelationshipTargets.add(relationshipTarget);
+      }
+      final AssessmentRelationshipDetail opponentRelationship =
+          new AssessmentRelationshipDetail()
+          .name(OPPONENT.getType().toLowerCase().replace("_", ""))
+          .relationshipTargets(opponentRelationshipTargets)
+          .prepopulated(true);
+
+      //proceeding relationships
+      final List<AssessmentRelationshipTargetDetail> proceedingRelationshipTargets =
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java:166]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentOpponentMappingContext.java (hunk 15 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/context/AssessmentOpponentMappingContext.java:16]: done

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java (hunk 18 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:983]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java (hunk 20 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:1047]: amended

File: src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java (hunk 22 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/service/AssessmentService.java:946]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+public class ApplicationUtil {
+
+  /**
+   * Gets the application type or amendment type for OPA input.
+   *
+   * @param application the application details
+   * @return "APP_TYPE_SUBSTANTIVE" if the application type is "EXCEPTIONAL_CASE_FUNDING",
+   *         otherwise the application type ID
+   */
+  public static String getAppAmendTypeOpaInput(final ApplicationDetail application) {
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java:25]: done

File: src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java (hunk 24 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/ApplicationUtil.java:56]: done

File: src/main/java/uk/gov/laa/ccms/caab/util/AssessmentUtil.java (hunk 26 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/AssessmentUtil.java:135]: done

File: src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java (hunk 27 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java:26]: done

File: src/main/java/uk/gov/laa/ccms/caab/util/ProceedingUtil.java (hunk 29 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/ProceedingUtil.java:26]: amended

File: src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java (hunk 28 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java:39]: amended

File: src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java [hunk 32]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  /**
+   * Constructs the full name of the given opponent based on its details and title lookup.
+   *
+   * @param opponent the opponent detail
+   * @param titleLookup the common lookup value detail for title descriptions
+   * @return the full name of the opponent, or "undefined" if no name is available
+   */
+  protected static String getFullName(
```
[the remark this answers is not in the document]
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/util/OpponentUtil.java:74]: Amended

File: src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java (hunk 13 above)
Comment (PhilDigitalJustice) [reply src/main/java/uk/gov/laa/ccms/caab/mapper/AssessmentMapper.java]: Amended all mapping methods

File: src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java [hunk 33]
```diff
@@ -0,0 +1,330 @@
...
+ * Represents a context token with various parameters required for validation and processing
+ * within the PUI and OWD (connector) systems.
+ */
+@Setter
+@Getter
+@Slf4j
+public class ContextToken {
+
+
+  // query String associate to this context
```
Comment (farrell-m) [inline src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java:24]:
```suggestion
  // query String associated to this context
```

File: src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java [hunk 34]
```diff
@@ -0,0 +1,330 @@
...
+  // i.e. {url}?val=xxxxxxx  used in checking the context within the PUI Connector Load() operation
+  // - e.g. TTL procesing 5-10 seconds.
```
Comment (farrell-m) [inline src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java:26]:
```suggestion
  // - e.g. TTL processing 5-10 seconds.
```

File: src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java [hunk 35]
```diff
@@ -0,0 +1,330 @@
...
+  public static final String QUERY_KEY_RET = "valret";
+
+  // common mandatory context param
+  // PUI business case ref.
+  public static final String CASE_ID = "id";
+  // rule base enumeration id
+  public static final String RULE_BASE_ID = "rb";
+  // pui Provider Id AKA OwnerId
+  public static final String PROVIDER_ID = "pid";
+  // timestamp An attempt to determine if thi context is valid within a time window -
```
Comment (farrell-m) [inline src/main/java/uk/gov/laa/ccms/caab/opa/context/ContextToken.java:39]:
```suggestion
  // timestamp An attempt to determine if this context is valid within a time window -
```

File: src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java [hunk 36]
```diff
@@ -0,0 +1,43 @@
+package uk.gov.laa.ccms.caab.opa.context.exception;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (farrell-m) [inline src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java:4]: Could be expanded to single imports

**Comment by farrell-m** [comment]
I assume the new variables will need to be added to the Helm Charts + AWS before this is merged?

File: src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java (hunk 36 above)
Comment (PhilDigitalJustice) [reply src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java:4]: I will raise a change for that

Comment (PhilDigitalJustice) [reply src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java:4]: https://github.com/ministryofjustice/laa-ccms-caab-helm-charts/pull/10

**Comment by PhilDigitalJustice** [comment]
> I assume the new variables will need to be added to the Helm Charts + AWS before this is merged?

https://github.com/ministryofjustice/laa-ccms-caab-helm-charts/pull/10

**Review by farrell-m — APPROVED** [review]
👍

File: src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java (hunk 36 above)
Comment (PhilDigitalJustice) [reply src/test/java/uk/gov/laa/ccms/caab/opa/context/exception/ConnectorSecurityContextExceptionTest.java:4]: Amended

**Pull request merged**